sdks/java/io/debezium/src/main/java/org/apache/beam/io/debezium/KafkaSourceConsumerFn.java

```diff
@@ -217,26 +239,44 @@ private static Instant ensureTimestampWithinBounds(Instant timestamp) {
   public ProcessContinuation process(
       @Element Map<String, String> element,
       RestrictionTracker<OffsetHolder, Map<String, Object>> tracker,
-      OutputReceiver<T> receiver)
-      throws Exception {
-    Map<String, String> configuration = new HashMap<>(element);
-
+      OutputReceiver<T> receiver) {
     // Adding the current restriction to the class object to be found by the database history
     register(tracker);
+    Map<String, String> configuration = new HashMap<>(element);
+    String cacheKey =
+        configuration.entrySet().stream()
+            .map(entry -> String.format("(%s,%s)", entry.getKey(), entry.getValue()))
+            .sorted()
+            .collect(Collectors.joining(","));
     configuration.put(BEAM_INSTANCE_PROPERTY, this.getHashCode());
+    SourceTask task;
 
-    SourceConnector connector = connectorClass.getDeclaredConstructor().newInstance();
-    connector.start(configuration);
-
-    SourceTask task = (SourceTask) connector.taskClass().getDeclaredConstructor().newInstance();
-
-    try {
+    if (CONNECTOR_CACHE.getIfPresent(cacheKey) == null) {
+      SourceConnector connector = null;
+      try {
+        connector = connectorClass.getDeclaredConstructor().newInstance();
+        connector.start(configuration);
+        task = (SourceTask) connector.taskClass().getDeclaredConstructor().newInstance();
+      } catch (InstantiationException
+          | IllegalAccessException
+          | InvocationTargetException
+          | NoSuchMethodException e) {
+        throw new RuntimeException("Unable to initialize connector instance for Debezium", e);
+      }
       Map<String, ?> consumerOffset = tracker.currentRestriction().offset;
-      LOG.debug("--------- Consumer offset from Debezium Tracker: {}", consumerOffset);
+      LOG.debug("--------- Created new Debezium task with offset: {}", consumerOffset);
 
       task.initialize(new BeamSourceTaskContext(tracker.currentRestriction().offset));
       task.start(connector.taskConfigs(1).get(0));
+      CONNECTOR_CACHE.put(cacheKey, task);
+    } else {
+      task = CONNECTOR_CACHE.getIfPresent(cacheKey);
```

Comment:
this may have been evicted from the cache since you first checked which is why using get with a callable is needed

Reply:
done